### DISCUSSION TITLE
Zähler SDM72D-M einrichten

### DISCUSSION THREAD

**Discussion by bebu259** [camueller/SmartApplianceEnabler#520] [Installation und Betrieb]
Hallo Axel,

ich hatte vor einiger Zeit einen `SDM72D-M` zum Ausprobieren erfolgreich angeschlossen und konnte ihn ausgelesen:

```
sae@raspberrypi:~ $ mbpoll -b 9600 -P none -a 1 -r 342 -t 3:float -B -0 -1 /dev/ttyUSB0
mbpoll 1.4-12 - FieldTalk(tm) Modbus(R) Master Simulator
Copyright © 2015-2019 Pascal JEAN, https://github.com/epsilonrt/mbpoll
This program comes with ABSOLUTELY NO WARRANTY.
This is free software, and you are welcome to redistribute it
under certain conditions; type 'mbpoll -w' for details.

Protocol configuration: Modbus RTU
Slave configuration...: address = [1]
                        start reference = 342, count = 1
Communication.........: /dev/ttyUSB0,       9600-8N1
                        t/o 1.00 s, poll rate 1000 ms
Data type.............: 32-bit float (big endian), input register table

-- Polling slave 1...
[342]: 	0.015
```

Jetzt ist der Zähler im Zählerschrank verbaut und ich bekomme beim Auslesen folgende Fehlermeldung:

```
sae@raspberrypi:~ $ mbpoll -b 9600 -P none -a 1 -r 342 -t 3:float -B -0 -1 /dev/ttyUSB0
terminate called after throwing an instance of 'std::system_error'
  what():  It seems that this system is not supported !: Operation not supported
Aborted
```

Am Raspi habe ich nach dem ersten Test eigentlich nichts geändert.
Die Polung (plus/minus) zum Zähler habe ich gecheckt, Widerstand ist auch an beiden Seiten vorhanden.

Der USB Adapter wird erkannt:
```
sae@raspberrypi:/ $ ls /dev/tty* | grep USB
/dev/ttyUSB0
```

Der Status von `mbusd` sieht so aus:
```
sae@raspberrypi:/ $ sudo systemctl status mbusd@ttyUSB0.service
[sudo] password for sae:
● mbusd@ttyUSB0.service - Modbus TCP to Modbus RTU (RS-232/485) gateway.
     Loaded: loaded (/lib/systemd/system/mbusd@.service; enabled; vendor preset: enabled)
     Active: inactive (dead) since Fri 2023-04-28 11:31:27 CEST; 22min ago
    Process: 1454 ExecStart=/usr/bin/mbusd -d -v2 -L - -c /etc/mbusd/mbusd-ttyUSB0.conf -p /dev/ttyUSB0 (code=exited, status=0/SUCCESS)
   Main PID: 1454 (code=exited, status=0/SUCCESS)
        CPU: 24ms

Apr 28 10:54:31 raspberrypi mbusd[1454]: 28 Apr 2023 10:54:31 tty_reopen(): can't open tty device /dev/ttyUSB0 (No such device)
Apr 28 10:54:32 raspberrypi mbusd[1454]: 28 Apr 2023 10:54:32 tty_reopen(): can't open tty device /dev/ttyUSB0 (No such file or directory)
Apr 28 10:54:32 raspberrypi mbusd[1454]: 28 Apr 2023 10:54:32 tty_reopen(): can't open tty device /dev/ttyUSB0 (No such file or directory)
Apr 28 10:54:32 raspberrypi mbusd[1454]: 28 Apr 2023 10:54:32 tty_reopen(): can't open tty device /dev/ttyUSB0 (No such file or directory)
Apr 28 10:54:33 raspberrypi mbusd[1454]: 28 Apr 2023 10:54:33 tty_reopen(): can't open tty device /dev/ttyUSB0 (No such file or directory)
Apr 28 10:54:35 raspberrypi mbusd[1454]: 28 Apr 2023 10:54:35 tty_reopen(): can't open tty device /dev/ttyUSB0 (No such file or directory)
Apr 28 11:31:27 raspberrypi mbusd[1454]: 28 Apr 2023 11:31:27 Terminated by signal: SIGTERM
Apr 28 11:31:27 raspberrypi systemd[1]: Stopping Modbus TCP to Modbus RTU (RS-232/485) gateway....
Apr 28 11:31:27 raspberrypi systemd[1]: mbusd@ttyUSB0.service: Succeeded.
Apr 28 11:31:27 raspberrypi systemd[1]: Stopped Modbus TCP to Modbus RTU (RS-232/485) gateway..
```

Stimmt doch, dass `mbusd` gestoppt sein muss, damit man `mbpoll` nutzen kann, oder?

Hast du eine Idee, woran es liegen könnte?

**Comment by camueller** [comment]
Du kannst den USB-Modbus-Adapter mal aus dem Raspberry Pi rausziehen und wieder reinstecken und dann mit dem Befehl `dmesg` schauen, ob der Adapter fehlerfrei initialisiert wird. Dabei wird das Gerät ja als `/dev/ttyUSB0` verfügbar gemacht.

**Comment by bebu259** [comment]
Dank, ich habe das mal versucht:

<img width="763" alt="Bildschirmfoto 2023-04-28 um 13 48 08" src="https://user-images.githubusercontent.com/1553063/235139936-721b8b19-b65c-4928-97b1-c6bb9ce90b28.png">

Der Adapter wird dann wieder erkannt:

<img width="370" alt="Bildschirmfoto 2023-04-28 um 13 48 52" src="https://user-images.githubusercontent.com/1553063/235139986-c3774d73-664b-4196-81e0-6a73f1b04408.png">

Aber irgendetwas ist mit `mbpoll` nicht in Ordnung, wenn ich nur `mbpoll -h` aufrufe, gibt es den selben Fehler:

<img width="657" alt="Bildschirmfoto 2023-04-28 um 13 51 17" src="https://user-images.githubusercontent.com/1553063/235140312-649f6993-75a2-475e-b453-d49d23ca9abe.png">

**Comment by camueller** [comment]
`mbpoll` soll ja nur ein Hilfsmittel sein, um zu prüfen, ob die Modbus-Kommunikation funktioniert. Die hat ja bei Dir schon mal funktioniert. Kann denn der SAE aktuell per Modbus auf den Zähler zugreifen?

**Reply by bebu259** [reply]
Ja, der SAE liest den Zähler korrekt aus, also funktioniert eigentlich alles.

Aber warum ich `mbpoll` nicht mehr aufrufen kann, ist mir rätselhaft.
Vielleicht hilft es, `mbpoll` einmal manuell neu zu installierten?

**Comment by camueller** [comment]
In der Vergangenheit hatte ich auch mal Probleme mit der `mbpoll`-Version, die über `apt` installiert wurde. Dann habe ich es mir aus den Source gebaut. Ist leicht und auf deren Homepage beschrieben: https://github.com/epsilonrt/mbpoll
